docs/using-the-compiler.rst

```diff
@@ -0,0 +1,273 @@
+find the file there, it will look at ``/usr/local/lib/fallback`` (the empty prefix
+always matches). ``solc`` will not read files from the filesystem that lie outside of
+the remapping targets and outside of the directories where explicitly specified source
+files reside, so things like ``import "/etc/passwd";`` only work if you add ``=/`` as a remapping.
+
+If there are multiple matches due to remappings, the one with the longest common prefix is selected.
+
+If your contracts use :ref:`libraries <libraries>`, you will notice that the bytecode contains substrings of the form ``__LibraryName______``. You can use ``solc`` as a linker meaning that it will insert the library addresses for you at those points:
+
+Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to your command to provide an address for each library or store the string in a file (one library per line) and run ``solc`` using ``--libraries fileName``.
+
+If ``solc`` is called with the option ``--link``, all input files are interpreted to be unlinked binaries (hex-encoded) in the ``__LibraryName____``-format given above and are linked in-place (if the input is read from stdin, it is written to stdout). All options except ``--libraries`` are ignored (including ``-o``) in this case.
+
+.. _compiler-api:
+
+Using the Compiler API
+**********************
+
+The compiler API expects a JSON formatted input and outputs the compilations result in a JSON formatted output.
+
+See the following pseudo-code:
+
+.. code-block:: none
+
+  // defined by the consumer of the API
+  importCallback(url:string) -> content:string
+
+  // invoking the compiler
+  solc.compile(inputJSON:string, importCallback:function) -> outputJSON:string
+
+The compiler will ask the ``importCallback`` for each URL defined for a source file and will stop when it succeeds.
+If all URLs failed, the compilation results in a failure.
+
+Compiler Input and Output JSON Description
+******************************************
+
+These JSON formats are used by the compiler API as well as are available through ``solc``. These are subject to change,
+some fields are optional (as noted), but it is aimed at to only make backwards compatible changes.
+
+Comments are of course not permitted and used here only for explanatory purposes.
+
+Input Description
+-----------------
+
+.. code-block:: none
+
+    {
+      sources:
+      {
+        // The keys here are the "global" names of the source files,
+        // imports can use other files via remappings (see below)
+        "myFile.sol":
+        {
+          // Optional: keccak256 hash of the source file
+          // It is used to verify the retrieved content if imported via URLs.
+          "keccak256": "0x123...",
+          // Required (unless "content" is used, see below): URL(s) to the source file.
+          "urls":
+          [
+            "bzzr://56ab...",
+            "ipfs://Qma...",
+            "file:///tmp/path/to/file.sol",
+          ]
```

Comment:
As per our meeting, decided to include `"stdin:"` here.

Now I don't think this is a good idea:
- if there are more entries specifying `stdin`, how do we read them?
- for stdin, the keccak256 hash cannot be verified

It doesn't really belong here. `solc`, if instructed to take the source from stdin, will just read it and provide it as a (literal source) `"content"` entry.